File: php/tests/Unit/Domain/Rules/Service/CompatibilityCheckerTest.php
<?php

declare(strict_types=1);

namespace Tests\Unit\Domain\Rules\Service;

use Alxarafe\App\Domain\Catalogue\ValueObject\ItemId;
use Alxarafe\App\Domain\Inventory\Entity\HandlingUnit;
use Alxarafe\App\Domain\Inventory\ValueObject\HandlingUnitId;
use Alxarafe\App\Domain\Inventory\ValueObject\HuStatus;
use Alxarafe\App\Domain\Inventory\ValueObject\Quantity;
use Alxarafe\App\Domain\Inventory\ValueObject\Sscc;
use Alxarafe\App\Domain\Inventory\ValueObject\StockQuantId;
use Alxarafe\App\Domain\Rules\Entity\CompatibilityRule;
use Alxarafe\App\Domain\Rules\Service\CompatibilityChecker;
use Alxarafe\App\Domain\Rules\ValueObject\AttributeId;
use Alxarafe\App\Domain\Rules\ValueObject\CompatibilityRuleId;
use Alxarafe\App\Domain\Rules\ValueObject\RuleScope;
use Alxarafe\App\Domain\Rules\ValueObject\RuleType;
use Alxarafe\App\Domain\Topology\Entity\Location;
use Alxarafe\App\Domain\Topology\ValueObject\AisleId;
use Alxarafe\App\Domain\Topology\ValueObject\LocationCode;
use Alxarafe\App\Domain\Topology\ValueObject\LocationId;
use Alxarafe\App\Domain\Topology\ValueObject\LocationRole;
use Alxarafe\App\Domain\Topology\ValueObject\LocationStatus;
use PHPUnit\Framework\TestCase;

final class CompatibilityCheckerTest extends TestCase
{
    private CompatibilityChecker $checker;
    private AttributeId $coldAttr;
    private AttributeId $frozenAttr;
    private AttributeId $refrigeratedAttr;
    private Location $activeLocation;
    private Location $inactiveLocation;

    protected function setUp(): void
    {
        $this->checker = new CompatibilityChecker();
        $this->coldAttr = new AttributeId('018e4e3a-3e7b-7b3e-8000-000000000001');
        $this->frozenAttr = new AttributeId('018e4e3a-3e7b-7b3e-8000-000000000002');
        $this->refrigeratedAttr = new AttributeId('018e4e3a-3e7b-7b3e-8000-000000000003');

        $aisleId = new AisleId('018e4e3a-3e7b-7b3e-8000-000000000010');
        $this->activeLocation = new Location(
            new LocationId('018e4e3a-3e7b-7b3e-8000-000000000011'),
            $aisleId,
            1,
            1,
            new LocationCode('WH01-A01-01-01'),
            LocationRole::PICKING,
            LocationStatus::ACTIVE,
        );
        $this->inactiveLocation = new Location(
            new LocationId('018e4e3a-3e7b-7b3e-8000-000000000012'),
            $aisleId,
            1,
            1,
            new LocationCode('WH01-A01-01-02'),
            LocationRole::PICKING,
            LocationStatus::BLOCKED,
        );
    }

    private function createHuWithItem(string $itemIdStr, ?string $quantIdStr = null): HandlingUnit
    {
        $huId = new HandlingUnitId('018e4e3a-3e7b-7b3e-8000-000000000020');
        $hu = new HandlingUnit($huId, new Sscc('123456789012345675'), null, null, HuStatus::AVAILABLE);
        $hu->addQuant(
            new StockQuantId($quantIdStr ?? '018e4e3a-3e7b-7b3e-8000-000000000030'),
            new ItemId($itemIdStr),
            null,
            new Quantity(1.0, 'EA'),
        );
        return $hu;
    }

    public function testInactiveLocationReturnsFalse(): void
    {
        $hu = $this->createHuWithItem('018e4e3a-3e7b-7b3e-8000-000000000100');
        $result = $this->checker->validate($hu, $this->inactiveLocation, [], [], []);
        self::assertFalse($result);
    }

    public function testEmptyHuReturnsTrue(): void
    {
        $hu = new HandlingUnit(
            new HandlingUnitId('018e4e3a-3e7b-7b3e-8000-000000000020'),
            new Sscc('123456789012345675'),
            null,
            null,
            HuStatus::AVAILABLE,
        );
        $result = $this->checker->validate($hu, $this->activeLocation, [], [], []);
        self::assertTrue($result);
    }

    public function testRequiresRuleSatisfied(): void
    {
        $hu = $this->createHuWithItem('018e4e3a-3e7b-7b3e-8000-000000000100');
        $itemFamilyAttrs = ['018e4e3a-3e7b-7b3e-8000-000000000100' => [$this->coldAttr]];
        $locationAttrs = [$this->refrigeratedAttr];
        $rules = [
            new CompatibilityRule(
                new CompatibilityRuleId('018e4e3a-3e7b-7b3e-8000-000000000050'),
                RuleType::REQUIRES,
                $this->coldAttr,
                $this->refrigeratedAttr,
                RuleScope::LOCATION,
            ),
        ];
        $result = $this->checker->validate($hu, $this->activeLocation, $itemFamilyAttrs, $locationAttrs, $rules);
        self::assertTrue($result);
    }

    public function testRequiresRuleNotSatisfied(): void
    {
        $hu = $this->createHuWithItem('018e4e3a-3e7b-7b3e-8000-000000000100');
        $itemFamilyAttrs = ['018e4e3a-3e7b-7b3e-8000-000000000100' => [$this->coldAttr]];
        $locationAttrs = [];
        $rules = [
            new CompatibilityRule(
                new CompatibilityRuleId('018e4e3a-3e7b-7b3e-8000-000000000050'),
                RuleType::REQUIRES,
                $this->coldAttr,
                $this->refrigeratedAttr,
                RuleScope::LOCATION,
            ),
        ];
        $result = $this->checker->validate($hu, $this->activeLocation, $itemFamilyAttrs, $locationAttrs, $rules);
        self::assertFalse($result);
    }

    public function testForbidsRuleNotViolated(): void
    {
        $hu = $this->createHuWithItem('018e4e3a-3e7b-7b3e-8000-000000000100');
        $itemFamilyAttrs = ['018e4e3a-3e7b-7b3e-8000-000000000100' => [$this->coldAttr]];
        $locationAttrs = [];
        $rules = [
            new CompatibilityRule(
                new CompatibilityRuleId('018e4e3a-3e7b-7b3e-8000-000000000050'),
                RuleType::FORBIDS,
                $this->coldAttr,
                $this->frozenAttr,
                RuleScope::LOCATION,
            ),
        ];
        $result = $this->checker->validate($hu, $this->activeLocation, $itemFamilyAttrs, $locationAttrs, $rules);
        self::assertTrue($result);
    }

    public function testForbidsRuleViolated(): void
    {
        $hu = $this->createHuWithItem('018e4e3a-3e7b-7b3e-8000-000000000100');
        $itemFamilyAttrs = ['018e4e3a-3e7b-7b3e-8000-000000000100' => [$this->coldAttr]];
        $locationAttrs = [$this->frozenAttr];
        $rules = [
            new CompatibilityRule(
                new CompatibilityRuleId('018e4e3a-3e7b-7b3e-8000-000000000050'),
                RuleType::FORBIDS,
                $this->coldAttr,
                $this->frozenAttr,
                RuleScope::LOCATION,
            ),
        ];
        $result = $this->checker->validate($hu, $this->activeLocation, $itemFamilyAttrs, $locationAttrs, $rules);
        self::assertFalse($result);
    }

    public function testMultipleItemsMultipleRules(): void
    {
        $hu = $this->createHuWithItem('018e4e3a-3e7b-7b3e-8000-000000000100');
        $itemId2 = '018e4e3a-3e7b-7b3e-8000-000000000101';
        $hu->addQuant(
            new StockQuantId('018e4e3a-3e7b-7b3e-8000-000000000031'),
            new ItemId($itemId2),
            null,
            new Quantity(2.0, 'EA'),
        );

        $itemFamilyAttrs = [
            '018e4e3a-3e7b-7b3e-8000-000000000100' => [$this->coldAttr],
            $itemId2 => [$this->frozenAttr],
        ];
        $locationAttrs = [$this->refrigeratedAttr];
        $rules = [
            new CompatibilityRule(
                new CompatibilityRuleId('018e4e3a-3e7b-7b3e-8000-000000000050'),
                RuleType::REQUIRES,
                $this->coldAttr,
                $this->refrigeratedAttr,
                RuleScope::LOCATION,
            ),
            new CompatibilityRule(
                new CompatibilityRuleId('018e4e3a-3e7b-7b3e-8000-000000000051'),
                RuleType::REQUIRES,
                $this->frozenAttr,
                $this->refrigeratedAttr,
                RuleScope::LOCATION,
            ),
        ];
        $result = $this->checker->validate($hu, $this->activeLocation, $itemFamilyAttrs, $locationAttrs, $rules);
        self::assertTrue($result);
    }
}

File: java/src/test/java/com/alxarafe/domain/rules/service/CompatibilityCheckerTest.java
package com.alxarafe.domain.rules.service;

import com.alxarafe.app.domain.catalogue.valueobject.ItemId;
import com.alxarafe.app.domain.inventory.entity.HandlingUnit;
import com.alxarafe.app.domain.inventory.valueobject.HandlingUnitId;
import com.alxarafe.app.domain.inventory.valueobject.HuStatus;
import com.alxarafe.app.domain.inventory.valueobject.Quantity;
import com.alxarafe.app.domain.inventory.valueobject.Sscc;
import com.alxarafe.app.domain.inventory.valueobject.StockQuantId;
import com.alxarafe.app.domain.rules.entity.CompatibilityRule;
import com.alxarafe.app.domain.rules.service.CompatibilityChecker;
import com.alxarafe.app.domain.rules.valueobject.AttributeId;
import com.alxarafe.app.domain.rules.valueobject.CompatibilityRuleId;
import com.alxarafe.app.domain.rules.valueobject.RuleScope;
import com.alxarafe.app.domain.rules.valueobject.RuleType;
import com.alxarafe.app.domain.topology.entity.Location;
import com.alxarafe.app.domain.topology.valueobject.AisleId;
import com.alxarafe.app.domain.topology.valueobject.LocationCode;
import com.alxarafe.app.domain.topology.valueobject.LocationId;
import com.alxarafe.app.domain.topology.valueobject.LocationRole;
import com.alxarafe.app.domain.topology.valueobject.LocationStatus;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.util.List;
import java.util.Map;

import static org.junit.jupiter.api.Assertions.*;

class CompatibilityCheckerTest {

    private CompatibilityChecker checker;
    private AttributeId coldAttr;
    private AttributeId frozenAttr;
    private AttributeId refrigeratedAttr;
    private Location activeLocation;
    private Location inactiveLocation;

    @BeforeEach
    void setUp() {
        checker = new CompatibilityChecker();
        coldAttr = new AttributeId("018e4e3a-3e7b-7b3e-8000-000000000001");
        frozenAttr = new AttributeId("018e4e3a-3e7b-7b3e-8000-000000000002");
        refrigeratedAttr = new AttributeId("018e4e3a-3e7b-7b3e-8000-000000000003");

        var aisleId = new AisleId("018e4e3a-3e7b-7b3e-8000-000000000010");
        activeLocation = new Location(
                new LocationId("018e4e3a-3e7b-7b3e-8000-000000000011"),
                aisleId, 1, 1,
                new LocationCode("WH01-A01-01-01"),
                LocationRole.PICKING,
                LocationStatus.ACTIVE);
        inactiveLocation = new Location(
                new LocationId("018e4e3a-3e7b-7b3e-8000-000000000012"),
                aisleId, 1, 1,
                new LocationCode("WH01-A01-01-02"),
                LocationRole.PICKING,
                LocationStatus.BLOCKED);
    }

    private HandlingUnit createHuWithItem(String itemIdStr, String quantIdStr) {
        var huId = new HandlingUnitId("018e4e3a-3e7b-7b3e-8000-000000000020");
        var hu = new HandlingUnit(huId, new Sscc("123456789012345675"), null, null, HuStatus.AVAILABLE);
        hu.addQuant(
                new StockQuantId(quantIdStr),
                new ItemId(itemIdStr),
                null,
                Quantity.fromDecimal(1.0, "EA"));
        return hu;
    }

    @Test
    void inactiveLocationReturnsFalse() {
        var hu = createHuWithItem("018e4e3a-3e7b-7b3e-8000-000000000100",
                "018e4e3a-3e7b-7b3e-8000-000000000030");
        assertFalse(checker.validate(hu, inactiveLocation, Map.of(), List.of(), List.of()));
    }

    @Test
    void emptyHuReturnsTrue() {
        var hu = new HandlingUnit(
                new HandlingUnitId("018e4e3a-3e7b-7b3e-8000-000000000020"),
                new Sscc("123456789012345675"),
                null, null, HuStatus.AVAILABLE);
        assertTrue(checker.validate(hu, activeLocation, Map.of(), List.of(), List.of()));
    }

    @Test
    void requiresRuleSatisfied() {
        var hu = createHuWithItem("018e4e3a-3e7b-7b3e-8000-000000000100",
                "018e4e3a-3e7b-7b3e-8000-000000000030");
        var itemFamilyAttrs = Map.of("018e4e3a-3e7b-7b3e-8000-000000000100", List.of(coldAttr));
        var locationAttrs = List.of(refrigeratedAttr);
        var rules = List.of(new CompatibilityRule(
                new CompatibilityRuleId("018e4e3a-3e7b-7b3e-8000-000000000050"),
                RuleType.REQUIRES, coldAttr, refrigeratedAttr, RuleScope.LOCATION));
        assertTrue(checker.validate(hu, activeLocation, itemFamilyAttrs, locationAttrs, rules));
    }

    @Test
    void requiresRuleNotSatisfied() {
        var hu = createHuWithItem("018e4e3a-3e7b-7b3e-8000-000000000100",
                "018e4e3a-3e7b-7b3e-8000-000000000030");
        var itemFamilyAttrs = Map.of("018e4e3a-3e7b-7b3e-8000-000000000100", List.of(coldAttr));
        var locationAttrs = List.<AttributeId>of();
        var rules = List.of(new CompatibilityRule(
                new CompatibilityRuleId("018e4e3a-3e7b-7b3e-8000-000000000050"),
                RuleType.REQUIRES, coldAttr, refrigeratedAttr, RuleScope.LOCATION));
        assertFalse(checker.validate(hu, activeLocation, itemFamilyAttrs, locationAttrs, rules));
    }

    @Test
    void forbidsRuleNotViolated() {
        var hu = createHuWithItem("018e4e3a-3e7b-7b3e-8000-000000000100",
                "018e4e3a-3e7b-7b3e-8000-000000000030");
        var itemFamilyAttrs = Map.of("018e4e3a-3e7b-7b3e-8000-000000000100", List.of(coldAttr));
        var locationAttrs = List.<AttributeId>of();
        var rules = List.of(new CompatibilityRule(
                new CompatibilityRuleId("018e4e3a-3e7b-7b3e-8000-000000000050"),
                RuleType.FORBIDS, coldAttr, frozenAttr, RuleScope.LOCATION));
        assertTrue(checker.validate(hu, activeLocation, itemFamilyAttrs, locationAttrs, rules));
    }

    @Test
    void forbidsRuleViolated() {
        var hu = createHuWithItem("018e4e3a-3e7b-7b3e-8000-000000000100",
                "018e4e3a-3e7b-7b3e-8000-000000000030");
        var itemFamilyAttrs = Map.of("018e4e3a-3e7b-7b3e-8000-000000000100", List.of(coldAttr));
        var locationAttrs = List.of(frozenAttr);
        var rules = List.of(new CompatibilityRule(
                new CompatibilityRuleId("018e4e3a-3e7b-7b3e-8000-000000000050"),
                RuleType.FORBIDS, coldAttr, frozenAttr, RuleScope.LOCATION));
        assertFalse(checker.validate(hu, activeLocation, itemFamilyAttrs, locationAttrs, rules));
    }

    @Test
    void multipleItemsMultipleRules() {
        var hu = createHuWithItem("018e4e3a-3e7b-7b3e-8000-000000000100",
                "018e4e3a-3e7b-7b3e-8000-000000000030");
        var itemId2 = "018e4e3a-3e7b-7b3e-8000-000000000101";
        hu.addQuant(
                new StockQuantId("018e4e3a-3e7b-7b3e-8000-000000000031"),
                new ItemId(itemId2), null, Quantity.fromDecimal(2.0, "EA"));

        var itemFamilyAttrs = Map.of(
                "018e4e3a-3e7b-7b3e-8000-000000000100", List.of(coldAttr),
                itemId2, List.of(frozenAttr));
        var locationAttrs = List.of(refrigeratedAttr);
        var rules = List.of(
                new CompatibilityRule(
                        new CompatibilityRuleId("018e4e3a-3e7b-7b3e-8000-000000000050"),
                        RuleType.REQUIRES, coldAttr, refrigeratedAttr, RuleScope.LOCATION),
                new CompatibilityRule(
                        new CompatibilityRuleId("018e4e3a-3e7b-7b3e-8000-000000000051"),
                        RuleType.REQUIRES, frozenAttr, refrigeratedAttr, RuleScope.LOCATION));
        assertTrue(checker.validate(hu, activeLocation, itemFamilyAttrs, locationAttrs, rules));
    }
}
